imgui_entt_entity_editor.hpp

```diff
@@ -1,200 +1,156 @@
+
+template <class Component, class EntityType>
+void ComponentAddAction(entt::basic_registry<EntityType>& registry, EntityType entity)
+{
+    registry.template assign<Component>(entity);
+}
+
+template <class Component, class EntityType>
+void ComponentRemoveAction(entt::basic_registry<EntityType>& registry, EntityType entity)
+{
+    registry.template remove<Component>(entity);
+}
+
+template <class EntityType>
+class EntityEditor {
+public:
+    using Registry = entt::basic_registry<EntityType>;
+
+    struct ComponentInfo {
+        using Callback = std::function<void(Registry&, EntityType)>;
+        std::string name;
+        Callback widget, create, destroy;
+    };
+
+    bool show_window = true;
+
+private:
+    using ComponentTypeID = ENTT_ID_TYPE;
+
+    std::map<ComponentTypeID, ComponentInfo> component_infos;
+
+    bool entityHasComponent(Registry& registry, EntityType& entity, ComponentTypeID type_id)
+    {
+        ComponentTypeID type[] = { type_id };
+        return registry.runtime_view(std::cbegin(type), std::cend(type)).contains(entity);
+    }
+
+public:
+    template <class Component>
+    ComponentInfo& registerComponent(const ComponentInfo& component_info)
+    {
+        auto index = entt::type_info<Component>::id();
+        auto [it, insert_result] = component_infos.insert_or_assign(index, component_info);
+        assert(insert_result);
+        auto& [type_index, info] = *it;
+        return info;
+    }
+
+    template <class Component>
+    ComponentInfo& registerComponent(const std::string& name, typename ComponentInfo::Callback widget)
+    {
+        return registerComponent<Component>(ComponentInfo{
+            name,
+            widget,
+            ComponentAddAction<Component, EntityType>,
+            ComponentRemoveAction<Component, EntityType>,
+        });
+    }
+
+    template <class Component>
+    ComponentInfo& registerComponent(const std::string& name)
+    {
+        return registerComponent<Component>(name, ComponentEditorWidget<Component, EntityType>);
+    }
+
+    void render(Registry& registry, EntityType& e)
+    {
+        if (show_window) {
+            if (ImGui::Begin("Entity Editor", &show_window)) {
+                ImGui::TextUnformatted("Editing:");
+                ImGui::SameLine();
+
+                if (registry.valid(e)) {
+                    ImGui::Text("ID: %d, Version: %d", static_cast<int>(registry.entity(e)), static_cast<int>(registry.version(e)));
+                } else {
+                    ImGui::Text("Invalid Entity");
+                }
+
+                if (ImGui::Button("New Entity")) {
+                    e = registry.create();
+                }
+
+                ImGui::Separator();
+
+                if (registry.valid(e)) {
+                    const std::string entity_id = std::to_string(static_cast<int>(registry.entity(e))) + std::to_string(static_cast<int>(registry.version(e)));
```

Comment:
I think `entt::to_integral(e)` would suffice.

Reply:
used to be `entt::to_integer()`

Reply:
Yes. That would indeed be better than the current method.